File: name-validation-3-python/REST/validate_name_rest.py
from nv3_response import NV3Response, NameResult, StatusDetail
import requests
import json

# Endpoint URLs for Service Objects Name Validation 3 (NV3) API
primary_url = "https://sws.serviceobjects.com/NV3/ValidateName?"
backup_url = "https://swsbackup.serviceobjects.com/NV3/ValidateName?"
trial_url = "https://trial.serviceobjects.com/NV3/ValidateName?"

def validate_name(
        full_name: str = "",
        prefix: str = "",
        first_name: str = "",
        middle_name: str = "",
        last_name: str = "",
        suffix: str = "",
        auth_id: str = "",
        is_live: bool = True,
        timeout_seconds: int = 15
) -> NV3Response:
    """
    Validates a name using the Service Objects Name Validation 3 (NV3) API. NV3 uses machine learning to
    to assess and classify input names. Will validate and score, as well as analyze for negative sentiment 
    and classify the input to distinguish between names and other types of inputs like business, dictionary, or garbage.
    Returns a StatusDetails response if invalid.


    Args:
        full_name (str, optional): The full name to validate. Required if no other name inputs are provided.
        prefix (str, optional): The name prefix (e.g., "Dr.", "Mr.").
        first_name (str, optional): The first name.
        middle_name (str, optional): The middle name.
        last_name (str, optional): The last name.
        suffix (str, optional): The name suffix (e.g., "Jr.", "III").
        auth_id (str): Your Service Objects authentication ID.
        is_live (bool, optional): Whether to use the live endpoint or the trial endpoint. Defaults to True (live).
        timeout_seconds (int, optional): Timeout for the API request in seconds.

    Returns:
        NV3Response: Parsed JSON response with name details or a StatusDetails if validation fails or the API call fails.

    Raises:
        RuntimeError: If the API call fails with a status of "500".
        requests.RequestException: On network/HTTP failures (trial mode).

    """

    params = {
        "FullName": full_name,
        "Prefix": prefix,
        "FirstName": first_name,
        "MiddleName": middle_name,
        "LastName": last_name,
        "Suffix": suffix,
        "AuthID": auth_id
    }

    url = primary_url if is_live else trial_url

    try:
        response = requests.get(url, params=params, timeout=timeout_seconds)
        response.raise_for_status()

        response_data = response.json()

        status_details = response_data.get("StatusDetail")
        if status_details:
            if is_live:
                #Try backup
                response = requests.get(backup_url, params=params, timeout=timeout_seconds)
                response.raise_for_status()
                status_details = response.json().get("StatusDetail")
                if status_details:
                    return NV3Response(StatusDetails=StatusDetail(**status_details))
            else:
                return NV3Response(StatusDetails=StatusDetail(**status_details))

        name_result = response_data.get("nameResult") or {}
        status_detail = response_data.get("statusDetail")

        return NV3Response(
            Status=response_data.get("status"),
            NameResult=NameResult(
                IsValidName=name_result.get("isValidName"),
                Classification=name_result.get("classification"),
                Confidence=name_result.get("confidence"),
                TextIn=name_result.get("textIn"),
                TextOut=name_result.get("textOut"),
                ParsedName=name_result.get("parsedName"),
                PossibleNames=name_result.get("possibleNames"),
                Notes=name_result.get("notes"),
                Warnings=name_result.get("warnings"),
                FirstNameFound=name_result.get("firstNameFound"),
                IsCommonFirstName=name_result.get("isCommonFirstName"),
                LastNameFound=name_result.get("lastNameFound"),
                IsCommonLastName=name_result.get("isCommonLastName"),
                SimilarFirstNames=name_result.get("similarFirstNames"),
                SimilarLastNames=name_result.get("similarLastNames"),
                RelatedNames=name_result.get("relatedNames"),
            ),
            StatusDetail=StatusDetail(
                Message=status_detail.get("message"),
                Detail=status_detail.get("detail"),
            ) if status_detail else None
        )
    except requests.RequestException as req_exc:
        if is_live:
            #Try backup
            try:
                response = requests.get(backup_url, params=params, timeout=timeout_seconds)
                response.raise_for_status()
                response_data = response.json()
                return NV3Response(
                    Status=response_data.get("status"),
                    NameResult=NameResult(
                        IsValidName=name_result.get("isValidName"),
                        Classification=name_result.get("classification"),
                        Confidence=name_result.get("confidence"),
                        TextIn=name_result.get("textIn"),
                        TextOut=name_result.get("textOut"),
                        ParsedName=name_result.get("parsedName"),
                        PossibleNames=name_result.get("possibleNames"),
                        Notes=name_result.get("notes"),
                        Warnings=name_result.get("warnings"),
                        FirstNameFound=name_result.get("firstNameFound"),
                        IsCommonFirstName=name_result.get("isCommonFirstName"),
                        LastNameFound=name_result.get("lastNameFound"),
                        IsCommonLastName=name_result.get("isCommonLastName"),
                        SimilarFirstNames=name_result.get("similarFirstNames"),
                        SimilarLastNames=name_result.get("similarLastNames"),
                        RelatedNames=name_result.get("relatedNames"),
                    ),
                    StatusDetail=StatusDetail(
                        Message=status_detail.get("message"),
                        Detail=status_detail.get("detail"),
                    ) if status_detail else None
                )
            except requests.RequestException as backup_exc:
                raise RuntimeError(f"API call failed for both primary and backup endpoints: {backup_exc}") from backup_exc

File: name-validation-3-nodejs/REST/validate_name_rest.js
import axios from 'axios';
import queryString from 'querystring';
import { NV3Response, NameResult, StatusDetail } from './nv3_response.js';

/**
 * @constant
 * @type {string}
 * @description Base URL for the live Service Objects Name Validation 3 (NV3) API service.
 */
const liveBaseUrl = 'https://sws.serviceobjects.com/NV3/';

/**
 * @constant
 * @type {string}
 * @description Base URL for the backup Service Objects Name Validation 3 (NV3) API service.
 */
const backupBaseUrl = 'https://swsbackup.serviceobjects.com/NV3/';

/**
 * @constant
 * @type {string}
 * @description Trial URL for the backup Service Objects Name Validation 3 (NV3) API service.
 */
const trialBaseUrl = 'https://trial.serviceobjects.com/NV3/';

/**
 * Checks if a response from the API is valid by verifying that it has no StatusDetail object
 * @param {NV3Response} response - The response object to check
 * @returns {boolean} - True if the response is valid, false otherwise
 */
const isValid = (response) => !response?.NameResult !==null;

/**
 * Constructs the full API URL by combining the base URL and URL-encoded query string parameters
 * derived from the input parameters.
 * @param {string} baseUrl - The base URL for the API endpoint (live, backup, or trial)
 * @param {Object} params - The query parameters for the API request
 * @returns {string} - The full API URL
 */
const buildUrl = (baseUrl, params) => 
    `${baseUrl}ValidateName?${queryString.stringify(params)}`;

/**
 * Performs an HTTP GET request to the specified URL with a given timeout.
 * @param {string} url - The URL to send the GET request to
 * @param {number} timeoutSeconds - The timeout for the request in seconds
 * @returns {Promise<NV3Response>} - A promise that resolves to an NV3Response object containing the API response data
 * @throws {Error} - Throws an error if the request fails
 */
export const httpGet = async (url, timeoutSeconds) => {
    let result = new NV3Response();
    try {
        const response = await axios.get(url, { timeout: timeoutSeconds * 1000 });

        if (response.status === 200) {
            result.Status = response.data.status ?? null;
            result.NameResult = response.data.nameResult ? new NameResult(response.data.nameResult) : null;
            result.StatusDetail = response.data.statusDetail ? new StatusDetail(response.data.statusDetail) : null;
        }
        else {
            result.Status = null;
            result.NameResult = null;
            result.StatusDetail = null;
        }
    } catch (error) {
        throw new Error(`HTTP GET request failed: ${error.message}`);
    }
    return result;
};

const ValidateNameClient = {
    /**
     * Provides functionality to call the Service Objects Name Validation 3 (NV3) API's ValidateName endpoint,
     * retrieving name validation and  classification details for a given name. Includes fallback to backup 
     * endpoint for reliability in live mode.
     * @param {string} [FullName] - The full name to validate. Optional to use instead of individual name components.
     * @param {string} [Prefix] - The name prefix to validate. Optional.
     * @param {string} [FirstName] - The first name to validate. Optional if FullName is provided.
     * @param {string} [MiddleName] - The middle name to validate. Optional.
     * @param {string} [LastName] - The last name to validate. Optional if FullName is provided.
     * @param {string} [Suffix] - The name suffix to validate. Optional.
     * @param {string} [Options] - Additional options for the API request. Optional.
     * @param {string} AuthID - The license key for authenticating with the API. Required.
     * @param {boolean} [isLive=true] - Whether to use the live API endpoint (true) or the trial endpoint (false).
     * @param {number} [timeoutSeconds=15] - The timeout for the API request in seconds. Optional, defaults to 15 seconds.
     * @returns {Promise<NV3Response>} A promise that resolves to a NV3Response object.
    */
    async invokeAsync(FullName, Prefix, FirstName, MiddleName, LastName, Suffix, Options, AuthID, isLive = true, timeoutSeconds = 15) {
        const params = {
            FullName,
            Prefix,
            FirstName,
            MiddleName,
            LastName,
            Suffix,
            Options,
            AuthID
        };

        const url = buildUrl(isLive ? liveBaseUrl : trialBaseUrl, params);
        let response = await httpGet(url, timeoutSeconds);

        if (isLive && !isValid(response)) {
            const fallbackUrl = buildUrl(backupBaseUrl, params);
            const fallbackResponse = await httpGet(fallbackUrl, timeoutSeconds);
            return fallbackResponse;
        }

        return response;
    },

    /**
     * Synchronously invokes the ValidateName API endpoint by wrapping the async call
     * and awaiting its result immediately. Note: This method should be used cautiously
     * in Node.js as it blocks the event loop.
     * @param {string} [FullName] - The full name to validate. Optional to use instead of individual name components.
     * @param {string} [Prefix] - The name prefix to validate. Optional.
     * @param {string} [FirstName] - The first name to validate. Optional if FullName is provided.
     * @param {string} [MiddleName] - The middle name to validate. Optional.
     * @param {string} [LastName] - The last name to validate. Optional if FullName is provided.
     * @param {string} [Suffix] - The name suffix to validate. Optional.
     * @param {string} [Options] - Additional options for the API request. Optional.
     * @param {string} AuthID - The license key for authenticating with the API. Required.
     * @param {boolean} [isLive=true] - Whether to use the live API endpoint (true) or the trial endpoint (false).
     * @param {number} [timeoutSeconds=15] - The timeout for the API request in seconds. Optional, defaults to 15 seconds.
     * @returns {NV3Response} - The response from the API call
     */
    invoke(FullName, Prefix, FirstName, MiddleName, LastName, Suffix, Options, AuthID, isLive = true, timeoutSeconds = 15) {
        return (async () => await this.invokeAsync(
            FullName, Prefix, FirstName, MiddleName, LastName, Suffix, Options, AuthID, isLive, timeoutSeconds
        ))();
    }
};

export { ValidateNameClient, NV3Response };